### PR TITLE
190004 - remove readme entry from catalog file

### PR DESCRIPTION
**Pull request by Nadeem14456**
This PR includes changes for 
[PBI 190003](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/190003)
[PBI 190004](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/190004)
- Added lifecycle block for virtual_network_subnet_id

### CODE REVIEW

**Review by amogh14893 — APPROVED** [review]
LGTM

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 1]
```diff
@@ -191,31 +191,38 @@ public static bool CheckFilesInDownloadedZip(string? downloadFolderPath, string
         /// This method is used to check the README.TXT
         /// </summary>
         /// <param name="downloadFolderPath">Sets the path of the folder where the required file is downloaded</param>
-        /// <param name="readMeSearchFilter">Sets the value of the Readme File type based on Config out of AVCS, BLANK or {Query}</param>
+        /// <param name="readMeSearchFilter">Sets the value of the Readme File type based on Config out of AVCS, BLANK, NONE or {Query}</param>
         /// <returns></returns>
         public static bool CheckReadMeInBessExchangeSet(string? downloadFolderPath, string? readMeSearchFilter)
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs:196]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 2]
```diff
@@ -413,5 +420,23 @@ public static bool VerifyPermitFile(string? downloadFolderPath, string? keyFileT
...
+            var catalogFileContent = File.ReadAllText(Path.Combine(downloadFolderPath!, testConfiguration.exchangeSetDetails.ExchangeSetEncRootFolder!, testConfiguration.exchangeSetDetails.ExchangeSetCatalogueFile!));
+            if (readMeSearchFilter!.Equals("NONE"))
+            {
+                catalogFileContent.Contains(testConfiguration.exchangeSetDetails.ExchangeReadMeFile!).Should().BeFalse();
+            }
+            else
+            {
+                catalogFileContent.Contains(testConfiguration.exchangeSetDetails.ExchangeReadMeFile!).Should().BeTrue();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 3]
```diff
@@ -315,11 +317,25 @@ private async Task<bool> CreateBessBatchAsync(string downloadPath, string fileEx
...
                 else if (configQueueMessage.Type.ToUpper().Equals(BessType.CHANGE.ToString()))
                 {
-                    batchType = Batch.BessChangeZipBatch;
+                    if (configQueueMessage.ReadMeSearchFilter.ToUpper().Equals("NONE"))
+                    {
+                        batchType = Batch.BessNoneReadmeBatch;
+                    }
+                    else
+                    {
+                        batchType = Batch.BessChangeZipBatch;
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService.UnitTests/Services/BuilderServiceTests.cs [hunk 4]
```diff
@@ -2005,6 +2010,227 @@ public async Task WhenTypeIsBase_AndNoLatestProductVersionFoundThenEmptyBespokeE
             ).MustHaveHappenedOnceExactly();
         }
 
+        [Test]
+        [TestCase(ExchangeSetStandard.S63)]
+        [TestCase(ExchangeSetStandard.S57)]
+        public async Task WhenTypeIsBaseAndReadmeSearchFilterIsNONE_ThenReadmeFileDeletedAndRemovedReadMeEntriesFromCatalogFileAndBespokeExchangeSetIsCreated(ExchangeSetStandard exchangeSetStandard)
```
[the remark this answers is not in the document]
Comment (tusha1501874) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService.UnitTests/Services/BuilderServiceTests.cs]: Resolved as per suggested.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common.UnitTests/Helpers/Catalog031ReaderFactoryWrapperTests.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        public void WhenGivenValidByteArray_ThenCreateShouldReturnCatalog031ReaderInstance()
+        {
+            var catalog031Reader = _catalog031ReaderFactoryWrapper.Create(_catalogFileBytes);
+            catalog031Reader.Should().BeOfType<Catalog031Reader>();
+        }
+
+        [Test]
+        public void WhenGivenNullByteArray_ThenCreateShouldThrowArgumentNullException()
+        {
+            Action act = () => _catalog031ReaderFactoryWrapper.Create(null);
```
[the remark this answers is not in the document]
Comment (tusha1501874) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common.UnitTests/Helpers/Catalog031ReaderFactoryWrapperTests.cs]: Resolved as per suggested.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/ICatalog031ReaderFactory.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿using UKHO.Torus.Enc.Core.EncCatalogue;
+
+namespace UKHO.PeriodicOutputService.Common.Helpers
+{
+    public interface ICatalog031ReaderFactory
+    {
+        ICatalog031Reader Create(byte[] catalogFileBytes);
```
[the remark this answers is not in the document]
Comment (tusha1501874) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/ICatalog031ReaderFactory.cs]: Resolved as per suggested.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Catalog031ReaderFactoryWrapper.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using UKHO.Torus.Enc.Core.EncCatalogue;
+
+namespace UKHO.PeriodicOutputService.Common.Helpers
+{
+    public class Catalog031ReaderFactoryWrapper : ICatalog031ReaderFactory
+    {
+        public ICatalog031Reader Create(byte[] catalogFileBytes)
```
[the remark this answers is not in the document]
Comment (tusha1501874) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Catalog031ReaderFactoryWrapper.cs]: Resolved as per suggested.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Catalog031Helper.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        private IEnumerable<CatalogEntry> GetCatalogEntries(string catalogFilePath)
+        {
+            byte[] catalogFileBytes = File.ReadAllBytes(catalogFilePath);
+            return _catalog031ReaderFactory.Create(catalogFileBytes).ReadCatalogue();
+        }
+
+        private byte[] FilterAndWriteCatalogEntries(IEnumerable<CatalogEntry> catalogEntries)
+        {
+            var catBuilder = _catalog031BuilderFactory.Create();
```
[the remark this answers is not in the document]
Comment (tusha1501874) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Catalog031Helper.cs]: Resolved as per suggested.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common.UnitTests/Helpers/Catalog031HelperTests.cs [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+            _catalog031Helper.RemoveReadmeEntryAndUpdateCatalogFile(_catalogFilePath);
+
+            VerifyCatalogEntriesAdded(catalogEntries);
+            VerifyCatalogFileOperations();
+            VerifyLoggingStart();
+            VerifyLoggingComplete();
+        }
+
+        [Test]
+        public void WhenCatalogContainsReadMeFile_ThenRemoveReadmeEntryAndUpdateCatalog_ShouldSkipsReadMeEntryAndUpdatesCatalog()
```
[the remark this answers is not in the document]
Comment (tusha1501874) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common.UnitTests/Helpers/Catalog031HelperTests.cs]: Resolved as per suggested.

**Review by Arit14975 — APPROVED** [review]
LGTM from testing side. Thanks

**Review by DanOwenUKHO — APPROVED** [review]
LGTM

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Pull request merged**